mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,662 @@
+        ),
+        axis_name="device",
+        donate_argnums=0,
+    )
+
+    return pmaped_update_step  # type:ignore
+
+
+def run_experiment(cfg: DictConfig) -> float:
+    # Add runtime variables to config
+    cfg.arch.n_devices = len(jax.devices())
+
+    # Number of env steps before evaluating/logging.
+    steps_per_rollout = int(cfg.system.total_timesteps // cfg.arch.num_evaluation)
+    # Multiplier for a single env/learn step in an anakin system
+    anakin_steps = cfg.arch.n_devices * cfg.system.update_batch_size
+    # Number of env steps in one anakin style update.
+    anakin_act_steps = anakin_steps * cfg.arch.num_envs * cfg.system.rollout_length
+    # Number of steps to do in the scanned update method (how many anakin steps).
+    cfg.system.scan_steps = int(steps_per_rollout / anakin_act_steps)
+
+    pprint(OmegaConf.to_container(cfg, resolve=True))
+
+    # Initialise system and make learning/evaluation functions
+    (env, eval_env), q_net, opt, rb, learner_state, logger, key = init(cfg)
+    update = make_update_fns(cfg, env, q_net, opt, rb)
+
+    cfg.system.num_agents = env.num_agents
+
+    key, eval_key = jax.random.split(key)
+    evaluator, absolute_metric_evaluator = make_eval_fns(
+        eval_env=eval_env,
+        network_apply_fn=q_net.apply,
+        config=cfg,
+        use_recurrent_net=True,
+        scanned_rnn=ScannedRNN(cfg.network.hidden_state_dim),
+    )
+
+    if cfg.logger.checkpointing.save_model:
+        checkpointer = Checkpointer(
+            metadata=cfg,  # Save all config as metadata in the checkpoint
+            model_name=cfg.logger.system_name,
+            **cfg.logger.checkpointing.save_args,  # Checkpoint args
+        )
+
+    max_episode_return = -jnp.inf
+    start_time = time.time()
+
+    # Main loop:
+    for eval_idx, t in enumerate(
+        range(steps_per_rollout, int(cfg.system.total_timesteps), steps_per_rollout)
+    ):
+        # Learn loop:
+        learner_state, (metrics, losses) = update(learner_state)
+        jax.block_until_ready(learner_state)
+
+        # Log:
+        # Add learn steps here because anakin steps per second is learn + act steps
+        # But we also want to make sure we're counting env steps correctly so
+        # learn steps is not included in the loop counter.
+        learn_steps = anakin_steps * cfg.system.epochs
+        eps = jax.numpy.maximum(
+            cfg.system.eps_min, 1 - (t / cfg.system.eps_decay) * (1 - cfg.system.eps_min)
+        )
+        sps = (t + learn_steps) / (time.time() - start_time)
+
+        final_metrics, ep_completed = episode_metrics.get_final_step_metrics(metrics)
+        loss_metrics = losses
+        logger.log({"step": t, "steps_per_second": sps}, t, eval_idx, LogEvent.MISC)
+        if ep_completed:
+            logger.log(final_metrics, t, eval_idx, LogEvent.ACT)
+        logger.log(loss_metrics, t, eval_idx, LogEvent.TRAIN)
+        logger.log({"epsilon": eps}, t, eval_idx, LogEvent.MISC)
```

Comment:
Just have 1 misc log event
```suggestion
        logger.log({"step": t, "steps_per_second": sps, "epsilon": eps}, t, eval_idx, LogEvent.MISC)
        if ep_completed:
            logger.log(final_metrics, t, eval_idx, LogEvent.ACT)
        logger.log(loss_metrics, t, eval_idx, LogEvent.TRAIN)
```